### PR TITLE
Integrate file download & thumbnail download into single fn

### PR DESCRIPTION
**Pull request by bitfriend**
Will integrate `get_avatar()` and `get_thumbnail()` into single fn

### CODE REVIEW

File: native/acter/src/api/room.rs [hunk 1]
```diff
@@ -1445,22 +1451,44 @@ impl Room {
...
+                    let available = matches!(
+                        &original.content.msgtype,
+                        MessageType::Image(_)
+                            | MessageType::Video(_)
+                            | MessageType::File(_)
+                            | MessageType::Location(_)
+                    );
+                    if !available {
+                        bail!("This message type is not downloadable");
+                    }
```
Comment (gnunicorn) [inline native/acter/src/api/room.rs:1465]: That's an odd way of writing this. Can't we directly do the if please?
```suggestion
                    if !matches!(
                        &original.content.msgtype,
                        MessageType::Image(_)
                            | MessageType::Video(_)
                            | MessageType::File(_)
                            | MessageType::Location(_)
                    ){
                        bail!("This message type is not downloadable");
                    }
```

File: native/acter/src/api/room.rs [hunk 2]
```diff
@@ -1445,22 +1451,44 @@ impl Room {
...
+                } else {
+                    let available = matches!(
```
Comment (gnunicorn) [inline native/acter/src/api/room.rs:1467]: and as above

**Review by gnunicorn — APPROVED** [review]
Looks good. aside from some minor rust style and an API fn location question

File: native/acter/api.rsh [hunk 3]
```diff
@@ -1810,7 +1799,7 @@ object Account {
     fn set_display_name(name: string) -> Future<Result<bool>>;
 
     /// The avatar of the client
-    fn avatar() -> Future<Result<OptionBuffer>>;
+    fn avatar(thumb_size: Option<ThumbnailSize>) -> Future<Result<OptionBuffer>>;
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/api.rsh:1818]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/53e71a13fc41d4eaa9de173002d05700c340a563

File: native/acter/api.rsh [hunk 4]
```diff
@@ -1496,7 +1485,7 @@ object SpaceHierarchyRoomInfo {
     fn has_avatar() -> bool;
 
     /// get the binary data of avatar
-    fn get_avatar() -> Future<Result<OptionBuffer>>;
+    fn get_avatar(thumb_size: Option<ThumbnailSize>) -> Future<Result<OptionBuffer>>;
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/api.rsh:1502]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/53e71a13fc41d4eaa9de173002d05700c340a563

File: native/acter/api.rsh [hunk 5]
```diff
@@ -1122,7 +1111,7 @@ object Attachment {
     /// if this is a media, hand over the description
     fn msg_content() -> MsgContent;
     /// if this is a media, hand over the data
-    fn source_binary() -> Future<Result<buffer<u8>>>;
+    fn source_binary(thumb_size: Option<ThumbnailSize>) -> Future<Result<buffer<u8>>>;
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/api.rsh:1126]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/53e71a13fc41d4eaa9de173002d05700c340a563

File: native/acter/api.rsh [hunk 6]
```diff
@@ -1007,19 +1003,12 @@ object Convo {
     /// get the users that were invited to this room
     fn get_invitees() -> Future<Result<Vec<Member>>>;
 
-    /// download media (image/audio/video/file) to specified path
-    fn download_media(event_id: string, dir_path: string) -> Future<Result<string>>;
+    /// download media (image/audio/video/file/location) to specified path
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/api.rsh:1014]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/53e71a13fc41d4eaa9de173002d05700c340a563

File: native/acter/api.rsh [hunk 7]
```diff
@@ -984,7 +980,7 @@ object Convo {
     /// decrypted media file data
     /// The reason that this function belongs to room object is because ChatScreen keeps it as member variable
     /// If this function belongs to message object, we may have to load too many message objects in ChatScreen
-    fn media_binary(event_id: string) -> Future<Result<buffer<u8>>>;
+    fn media_binary(event_id: string, thumb_size: Option<ThumbnailSize>) -> Future<Result<buffer<u8>>>;
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/api.rsh:991]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/53e71a13fc41d4eaa9de173002d05700c340a563

File: native/acter/api.rsh [hunk 8]
```diff
@@ -293,7 +289,7 @@ object NewsSlide {
     /// if this is a media, hand over the description
     fn msg_content() -> MsgContent;
     /// if this is a media, hand over the data
-    fn source_binary() -> Future<Result<buffer<u8>>>;
+    fn source_binary(thumb_size: Option<ThumbnailSize>) -> Future<Result<buffer<u8>>>;
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/api.rsh:298]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/53e71a13fc41d4eaa9de173002d05700c340a563

File: native/acter/api.rsh [hunk 9]
```diff
@@ -168,10 +165,7 @@ object RoomProfile {
     fn has_avatar() -> Result<bool>;
 
     /// get the binary data of avatar
-    fn get_avatar() -> Future<Result<OptionBuffer>>;
-
-    /// get the binary data of thumbnail
-    fn get_thumbnail(width: u64, height: u64) -> Future<Result<OptionBuffer>>;
+    fn get_avatar(thumb_size: Option<ThumbnailSize>) -> Future<Result<OptionBuffer>>;
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/api.rsh:172]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/53e71a13fc41d4eaa9de173002d05700c340a563

File: native/acter/api.rsh [hunk 10]
```diff
@@ -154,10 +154,7 @@ object UserProfile {
     fn has_avatar() -> Future<Result<bool>>;
 
     /// get the binary data of avatar
-    fn get_avatar() -> Future<Result<OptionBuffer>>;
-
-    /// get the binary data of thumbnail
-    fn get_thumbnail(width: u64, height: u64) -> Future<Result<OptionBuffer>>;
+    fn get_avatar(thumb_size: Option<ThumbnailSize>) -> Future<Result<OptionBuffer>>;
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/api.rsh:159]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/53e71a13fc41d4eaa9de173002d05700c340a563

File: native/acter/api.rsh [hunk 11]
```diff
@@ -2096,6 +2085,9 @@ object Client {
     /// upload file and return remote url
     fn upload_media(uri: string) -> Future<Result<MxcUri>>;
 
+    /// create size object to be used for thumbnail download
+    fn new_thumb_size(width: u64, height: u64) -> Result<ThumbnailSize>;
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/api.rsh]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/df02c722ff6c27a8abef86fe98877fa7cb20f14f

File: native/acter/src/api/attachments.rs [hunk 12]
```diff
@@ -87,25 +89,88 @@ impl Attachment {
...
-            }
+            AttachmentContent::Image(content) => match thumb_size {
+                Some(thumb_size) => {
+                    let source = content
+                        .info
+                        .as_ref()
+                        .and_then(|info| info.thumbnail_source.clone());
+                    let Some(source) = source else {
+                        bail!("thumbnail source doesn't exist")
+                    };
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/src/api/attachments.rs]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/079297324b9de418355af6c8248aa47e5aa573c8

File: native/acter/src/api/attachments.rs [hunk 13]
```diff
@@ -87,25 +89,88 @@ impl Attachment {
...
+                        .await
+                }
+            },
+            AttachmentContent::Video(content) => match thumb_size {
+                Some(thumb_size) => {
+                    let source = content
+                        .info
+                        .as_ref()
+                        .and_then(|info| info.thumbnail_source.clone());
+                    let Some(source) = source else {
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/src/api/attachments.rs]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/079297324b9de418355af6c8248aa47e5aa573c8

File: native/acter/src/api/news.rs [hunk 14]
```diff
@@ -210,21 +214,92 @@ impl NewsSlide {
...
-                let buf = Vec::<u8>::new();
-                Ok(FfiBuffer::new(buf))
-            }
+            NewsContent::Image(content) => match thumb_size {
+                Some(thumb_size) => {
+                    let source = content
+                        .info
+                        .as_ref()
+                        .and_then(|info| info.thumbnail_source.clone());
+                    let Some(source) = source else {
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/src/api/news.rs]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/079297324b9de418355af6c8248aa47e5aa573c8

File: native/acter/src/api/attachments.rs [hunk 15]
```diff
@@ -87,25 +89,88 @@ impl Attachment {
...
+            AttachmentContent::Audio(content) => match thumb_size {
+                Some(thumb_size) => {
+                    bail!("audio has not thumbnail")
+                }
+                None => {
+                    self.client
+                        .source_binary(content.source.clone(), None)
+                        .await
+                }
+            },
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/src/api/attachments.rs]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/3d4bbf7495f59f482079236c845484b675613166

File: native/acter/src/api/client.rs [hunk 16]
```diff
@@ -375,13 +378,22 @@ impl Client {
...
+        source: MediaSource,
+        thumb_size: Option<Box<ThumbnailSize>>,
+    ) -> Result<FfiBuffer<u8>> {
         // any variable in self can't be called directly in spawn
         let client = self.clone();
-        let request = MediaRequest {
-            source,
-            format: MediaFormat::File,
+        let format = match thumb_size {
+            Some(thumb_size) => MediaFormat::Thumbnail(MediaThumbnailSize {
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/src/api/client.rs]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/903d20a73be4a6b0dec2fe47131eb7eb23552b8d

File: native/acter/src/api/room.rs (hunk 1 above)
Comment (bitfriend) [reply native/acter/src/api/room.rs:1429]: When `if` case becomes simple, github workflow requires `else` case to be changed into `else if`.
It looks like the following code snippet.
```rs
                if is_thumb
                    && !matches!(
                        &original.content.msgtype,
                        MessageType::Image(_)
                            | MessageType::Video(_)
                            | MessageType::File(_)
                            | MessageType::Location(_)
                    ) {
                    bail!("This message type is not downloadable");
                } else if !matches!(
                    &original.content.msgtype,
                    MessageType::Image(_)
                        | MessageType::Audio(_)
                        | MessageType::Video(_)
                        | MessageType::File(_)
                ) {
                    bail!("This message type is not downloadable");
                }
```
As you see, 1st block seems to have check of `is_thumb` more than 2nd block.
I think this code may let developer misunderstand easily.

Comment (bitfriend) [reply native/acter/src/api/room.rs:1429]: So I placed the variable of `available` in order to give developer clear understanding and block failure in github workflow lint.

File: native/acter/src/api/common.rs [hunk 17]
```diff
@@ -544,3 +547,18 @@ impl ThumbnailSize {
...
+pub(crate) fn into_media_format(thumb_size: Option<Box<ThumbnailSize>>) -> MediaFormat {
+    match thumb_size {
+        Some(thumb_size) => MediaFormat::Thumbnail(MediaThumbnailSize {
+            method: get_content_thumbnail::v3::Method::Scale,
+            width: thumb_size.width(),
+            height: thumb_size.height(),
+        }),
+        None => MediaFormat::File,
+    }
+}
```
Comment (gnunicorn) [inline native/acter/src/api/common.rs:564]: let's move this onto `ThumbnailSize` and split it up please:

```suggestion
impl ThumbnailSize {
   pub fn into_media_format(&self) -> MediaFormat {
       MediaFormat::Thumbnail(MediaThumbnailSize {
            method: get_content_thumbnail::v3::Method::Scale,
            width: self.width(),
            height: self.height(),
        }
   }
   
   pub fn parse_into_media_format(thumb_size: Option<Box<ThumbnailSize>>) -> MediaFormat {
        match thumb_size {
            Some(thumb_size) => thumb_size.into_media_format(),
            None => MediaFormat::File,
       }
    }
   
```

and then use `ThumbnailSize::parse_into_media_format(a)`  and `thumbnail_size.into_media_format()` for calling it

File: native/acter/src/api/profile.rs [hunk 18]
```diff
@@ -179,14 +172,7 @@ impl RoomProfile {
...
             .context("Room not found")?;
-        let format = match thumb_size {
-            Some(thumb_size) => MediaFormat::Thumbnail(MediaThumbnailSize {
-                method: get_content_thumbnail::v3::Method::Scale,
-                width: thumb_size.width(),
-                height: thumb_size.height(),
-            }),
-            None => MediaFormat::File,
-        };
+        let format = into_media_format(thumb_size);
```
Comment (gnunicorn) [inline native/acter/src/api/profile.rs:175]: As per the API change suggested above 
```suggestion
        let format = ThumbnailSize::parse_into_media_format(thumb_size);
```

File: native/acter/src/api/room.rs [hunk 19]
```diff
@@ -1243,18 +1225,12 @@ impl Room {
                             (request, filename)
                         }
                         MessageType::Video(content) => {
+                            let format = into_media_format(Some(thumb_size));
```
Comment (gnunicorn) [inline native/acter/src/api/room.rs:1228]: Suggestions above makes this smaller and only if actually a source is found by doing:
```suggestion
```

File: native/acter/src/api/room.rs [hunk 20]
```diff
@@ -1243,18 +1225,12 @@ impl Room {
...
                                 .and_then(|info| info.thumbnail_source.clone())
-                                .map(|source| MediaRequest {
-                                    source,
-                                    format: MediaFormat::Thumbnail(MediaThumbnailSize {
-                                        method: get_content_thumbnail::v3::Method::Scale,
-                                        width: thumb_size.width(),
-                                        height: thumb_size.height(),
-                                    }),
-                                });
+                                .map(|source| MediaRequest { source, format });
```
Comment (gnunicorn) [inline native/acter/src/api/room.rs:1233]:
```suggestion
                                .map(|source| MediaRequest { source, format: thumb_size.into() });
```

**Review by gnunicorn — APPROVED** [review]
Very nice, thank you!

Just one more minor type API suggestion where and how to plug the specific implementation. I only commented a single instance of each, but I think it is clear how this is meant to be used everywhere.

File: native/acter/src/api/common.rs (hunk 17 above)
Comment (bitfriend) [reply native/acter/src/api/common.rs]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/518193fa236e942607e738383ba8ffb64b96b9ca

File: native/acter/src/api/room.rs (hunk 19 above)
Comment (bitfriend) [reply native/acter/src/api/room.rs]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/518193fa236e942607e738383ba8ffb64b96b9ca

File: native/acter/src/api/room.rs (hunk 20 above)
Comment (bitfriend) [reply native/acter/src/api/room.rs]: Fixed by https://github.com/acterglobal/a3/pull/1246/commits/518193fa236e942607e738383ba8ffb64b96b9ca

**Pull request merged**